### PR TITLE
Add, Remove and Improve ebook(Fix issue:#2990, #2992, #2993, #2997, #2998)

### PR DESCRIPTION
**Pull request by bluetata**
## What does this PR do?
Add Resource(s) | Remove Resource(s) | Improve Repo

### Description 
Add, Remove and Improve ebook(Fix issue:#2990, #2992, #2993, #2997, #2998).

Fix Issue: #2990, replace book link to available new link.
Fix Issue: #2992, replace incorrect link to available new link. 
Fix Issue: #2993, add book: the black art of programming by mcilroy mark
Fix Issue: #2997, add book: A New Kind of Science by Stephen Wolfram
Fix Issue: #2998, add book: Foundations of Computer Science by Lawrence C Paulson

### Checklist:
- [x] Not a duplicate
- [x] Included author(s) if appropriate
- [x] Lists are in alphabetical order
- [x] Needed indications added (PDF, access notes, under construction)

### CODE REVIEW

File: free-programming-books.md [hunk 1]
```diff
@@ -360,8 +360,10 @@
 
 #### Computer Science
 
+* [A New Kind of Science](http://www.wolframscience.com/nks/) - Stephen Wolfram
```
Comment (eshellman) [inline free-programming-books.md]: I think this is a duplicate- under cellular autonomata

File: free-programming-books.md [hunk 2]
```diff
@@ -360,8 +360,10 @@
...
 * [Computer Science I - Draft](https://bitbucket.org/chrisbourke/computersciencei/src) - Dr. Chris Bourke (PDF) (:construction: *in process*)
 * [Computer Science II - Draft](http://cse.unl.edu/~cbourke/ComputerScienceTwo.pdf) - Dr. Chris Bourke (PDF) (:construction: *in process*)
+* [Foundations of Computer Science](http://www.cl.cam.ac.uk/teaching/2000/FoundsCS/Founds-FP.pdf) - Lawrence C Paulson (PDF)
```
Comment (eshellman) [inline free-programming-books.md]: There's another book with the same title under Algorithms & Data Structures

Comment (eshellman) [reply free-programming-books.md]: Also, this book describes itself as not-a-book lecture notes. If we think it still stands alone, we should use the full titel to distinguish it from the author's book.

File: free-programming-books.md [hunk 3]
```diff
@@ -1418,15 +1421,15 @@ Kerridge (PDF) (email address *requested*, not required)
 * [Java Language and Virtual Machine Specifications](https://docs.oracle.com/javase/specs/) - James Gosling, et al
 * [Java Notes for Professionals](http://books.goalkicker.com/JavaBook/) - Compiled from StackOverflow documentation (3.x)
 * [Java Programming for Kids](https://yfain.github.io/Java4Kids/) - Yakov Fain
-* [Java: The Legend](https://www.oreilly.com/programming/free/java-the-legend.csp) - Ben Evans (email address *requested*, not required)
+* [Java: The Legend](http://pepa.holla.cz/wp-content/uploads/2016/10/java-the-legend.pdf) - Ben Evans (PDF)
```
Comment (eshellman) [inline free-programming-books.md]: This doesn't appear to be an authorized download.

File: free-programming-books.md [hunk 4]
```diff
@@ -1418,15 +1421,15 @@ Kerridge (PDF) (email address *requested*, not required)
...
 * [Modern Java EE Design Patterns](http://www.oreilly.com/programming/free/modern-java-ee-design-patterns.csp) - Markus Eisele (email address *requested*, not required)
 * [Neural Network Development with Java](https://www.packtpub.com/packt/free-ebook/neural-networks-java) - Alan M. F. Souza and Fabio M. Soares (email address *requested*, not required)
 * [Object-Oriented Programming in JavaTM Textbook](http://computing.southern.edu/halterman/OOPJ/) - Rick Halterman (PDF per Chapter)
 * [Object Oriented Programming using Java](https://bookboon.com/en/object-oriented-programming-using-java-ebook) - Simon Kendal (PDF) , Bookboon. (email address *requested*, not required)
-* [Object-Oriented vs. Functional Programming](http://www.oreilly.com/programming/free/object-oriented-vs-functional-programming.csp) - Richard Warburton (email address *requested*, not required)
+* [Object-Oriented vs. Functional Programming](http://www.r-5.org/files/books/computers/languages/java/scala/Richard_Warburton-Object-Oriented_vs_Functional_Programming-EN.pdf) - Richard Warburton (PDF)
```
Comment (eshellman) [inline free-programming-books.md]: doesn't appear to be authorized

File: free-programming-books.md (hunk 1 above)
Comment (bluetata) [reply free-programming-books.md]: yes,  it's a duplicate

File: free-programming-books.md (hunk 2 above)
Comment (bluetata) [reply free-programming-books.md]: I think we need to discuss more for it in Issue.

File: free-programming-books.md (hunk 3 above)
Comment (bluetata) [reply free-programming-books.md]: I think we need to discuss more for it in Issue.

File: free-programming-books.md (hunk 4 above)
Comment (bluetata) [reply free-programming-books.md]: I think we need to discuss more for it in Issue.

**Comment by bluetata** [comment]
For those RR we will closed it and we need to discuss more for it in Issue list, and I will open a new PR for #2993 firstly

**Comment by bluetata** [comment]
Thanks very much for @eshellman 's review. and I will be careful and conscientious in the PR and issue check

**Pull request closed without merging**